### ISSUE TITLE
e2e flake: tests are failing with "net/http: TLS handshake timeout" in large clusters

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#26519]
We are running a standard e2e suite that we are running in gce-e2e in large 2000-node clusters now.

And on every run, we have few random tests failing with "net/http: TLS handshake timeout" error.

Examples:
```
[k8s.io] Pods
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:641
  should be restarted with a /healthz http liveness probe [Conformance] [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:1100

  getting pod liveness-http
  Expected error:
      <*url.Error | 0xc820e55f20>: {
          Op: "Get",
          URL: "https://104.196.136.191/api/v1/namespaces/e2e-tests-pods-uvjef/pods/liveness-http",
          Err: {},
      }
      Get https://104.196.136.191/api/v1/namespaces/e2e-tests-pods-uvjef/pods/liveness-http: net/http: TLS handshake timeout
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:86
```

or

```
• Failure in Spec Setup (BeforeEach) [61.291 seconds]
[k8s.io] ConfigMap
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:641
  should be consumable from pods in volume as non-root [Conformance] [BeforeEach]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/configmap.go:41

  Expected error:
      <*url.Error | 0xc82271f4a0>: {
          Op: "Get",
          URL: "https://104.196.136.191/api/v1/watch/namespaces/e2e-tests-configmap-b06fw/serviceaccounts?fieldSelector=metadata.name%3Ddefault",
          Err: {},
      }
      Get https://104.196.136.191/api/v1/watch/namespaces/e2e-tests-configmap-b06fw/serviceaccounts?fieldSelector=metadata.name%3Ddefault: net/http: TLS handshake timeout
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:169
```

@zmerlynn @gmarek @lavalamp @kubernetes/sig-api-machinery

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
BTW - we are far from being cpu-starved in those tests, so the issue is not lack of CPU (at least not at the master machine)

**Comment by wojtek-t** [comment]
Also, in my opinion https://github.com/kubernetes/kubernetes/issues/15664 seems to be unrelated.= here.

**Comment by smarterclayton** [comment]
Did we turn on http2?  I thought there was a bug I saw recently in Go 1.6
about HTTP2 and not consuming an entire stream if you didn't read all the
way AND close() (or similar), which leads to more connections than normal.

On May 30, 2016, at 9:17 AM, Wojciech Tyczynski <notifications@github.com>
wrote:

Also, in my opinion #15664
<https://github.com/kubernetes/kubernetes/issues/15664> seems to be
unrelated.= here.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/26519#issuecomment-222490742>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p6drtBHm3omaTVED5YxRiFTYYHF2ks5qGuNigaJpZM4Ipvox>
.

**Comment by wojtek-t** [comment]
Hmm - I think we didn't, but I'm not 100%.
@krousey @timothysc - did we enable http2 by default?

**Comment by dims** [comment]
@wojtek-t looks like the http2 PR is still in "Open" state

**Comment by wojtek-t** [comment]
@dims - that one I know, but the question is whether Go 1.6 doesn't enable something features of it by default. That's something I'm not sure about. @krousey ?

**Comment by wojtek-t** [comment]
In particular, from release notes: https://golang.org/doc/go1.6
```
Go 1.6 adds transparent support in the net/http package for the new HTTP/2 protocol. Go clients and servers will automatically use HTTP/2 as appropriate when using HTTPS. There is no exported API specific to details of the HTTP/2 protocol handling, just as there is no exported API specific to HTTP/1.1.

Programs that must disable HTTP/2 can do so by setting Transport.TLSNextProto (for clients) or Server.TLSNextProto (for servers) to a non-nil, empty map.

Programs that must adjust HTTP/2 protocol-specific details can import and use golang.org/x/net/http2, in particular its ConfigureServer and ConfigureTransport functions.
```

**Comment by smarterclayton** [comment]
I've seen bugs in the http2 stack around Close and http 1.0 No content
responses, or chunked encoding.  So it's possible that we're doing a lot
more tls handshakes than we expect, although I discount that somewhat
because I would expect pegged CPU on the master then.

On May 30, 2016, at 1:27 PM, Wojciech Tyczynski <notifications@github.com>
wrote:

In particular, from release notes: https://golang.org/doc/go1.6

Go 1.6 adds transparent support in the net/http package for the new
HTTP/2 protocol. Go clients and servers will automatically use HTTP/2
as appropriate when using HTTPS. There is no exported API specific to
details of the HTTP/2 protocol handling, just as there is no exported
API specific to HTTP/1.1.

Programs that must disable HTTP/2 can do so by setting
Transport.TLSNextProto (for clients) or Server.TLSNextProto (for
servers) to a non-nil, empty map.

Programs that must adjust HTTP/2 protocol-specific details can import
and use golang.org/x/net/http2, in particular its ConfigureServer and
ConfigureTransport functions.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/26519#issuecomment-222531184>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p_P0qwzRDjlVsI784lDCGRsbmdMuks5qGx3jgaJpZM4Ipvox>
.

**Comment by timothysc** [comment]
The servers will have it enabled, but none of the clients will by default.  

I've also changed my pull to default off and only option enable.

**Comment by wojtek-t** [comment]
@timothysc - thanks!

**Comment by wojtek-t** [comment]
@lavalamp @smarterclayton - do you think it makes sense to retry "handshake timeout" errors at the client layer? Exactly the same way we are retrying "Please try again later" errors?

**Comment by wojtek-t** [comment]
Actually - now I think that this is related to crashing apiserver. Time when tests fail matches times of apiserver crashes.
I will file another issue.

**Comment by wojtek-t** [comment]
Though, I don't have full logs to check it now. So I'm leaving this issue still open. But we definitely need to fix #26563 first.

**Comment by wojtek-t** [comment]
So it seems that it was false alarm. Those errors were because of unreachable apiserver.
And apiserver was unreachable, because it was crashing due to #26563 
After fixing that bug, I no longer see that issues (before fixing them I've seen them in almost every run of large clusters, now we had 10+ runs without it).

So I'm closing this issue.
